hebiros_gazebo_plugin/plugin_base/hebi_gazebo_plugin.cpp

```diff
@@ -0,0 +1,151 @@
+    res *= 10;
+    res += static_cast<int>(string_ver[i] - '0');
+    ++i;
+  }
+  return res;
+}
+
+// This is a templated struct that allows for wrapping some of the Gazebo
+// code for which compilation differse between versions; we use partial
+// template specialization to compile the appropriate version.
+template<int GazeboVersion, typename JointType> struct GazeboHelper {
+  static_assert(GazeboVersion == 7 || GazeboVersion == 9, "Unknown version of gazebo");
+  // Default implementations so that the above assertion is the only compilation error
+  static double position(JointType joint) { return 0; }
+  static double effort(JointType joint) { return 0; }
+};
+
+using GazeboWrapper = GazeboHelper<GetGazeboVersion(GAZEBO_VERSION), gazebo::physics::JointPtr>;
+
+template<typename JointType> struct GazeboHelper<7, JointType> {
+  static double position(JointType joint) {
+    return joint->GetAngle(0).Radian(); 
+  }
+
+  static double effort(JointType joint) {
+    auto trans = joint->GetChild()->GetInitialRelativePose().rot;
+    auto wrench = joint->GetForceTorque(0);
+    return (-1 * (trans * wrench.body1Torque)).z;
+  }
+};
+
+template<typename JointType> struct GazeboHelper<9, JointType> {
+  static double position(JointType joint) {
+    return joint->Position(0);
+  }
+
+  static double effort(JointType joint) {
+    auto trans = joint->GetChild()->InitialRelativePose().Rot();
+    gazebo::physics::JointWrench wrench = joint->GetForceTorque(0);
+    return (-1 * (trans * wrench.body1Torque)).Z();
+  }
+};
+
+namespace hebi {
+namespace sim {
+namespace plugin {
+
+void HebiGazeboPlugin::Load(gazebo::physics::ModelPtr model, sdf::ElementPtr sdf) {
+  model_ = model;
+
+  // Go through each joint in the model, looking for anything matching a
+  // <family>/<name>/<type> naming pattern, where <type> is a recognized
+  // module type.
+  // We add references to the matching joints so we can simulate them.
+  for (auto j : model_->GetJoints())
+  {
+    auto full_name = j->GetName();
+    auto first_slash = full_name.find("/");
+    // No slash!
+    if (first_slash == std::string::npos)
+      continue;
+    auto second_slash = full_name.find("/", first_slash + 1);
+    if (second_slash == std::string::npos)
+      continue;
+    
+    // Get type:
+    auto family = full_name.substr(0, first_slash);
+    auto name = full_name.substr(first_slash + 1, second_slash - first_slash - 1);
+    auto type = full_name.substr(second_slash + 1);
+
+    auto hebi_joint = Joint::tryCreate(family, name, type);
+
+    if (hebi_joint)
+    {
+      std::cout << "Adding parsed joint " << family << "/" << name << " of type: " << type << "\n";
+      joints_.push_back({ std::move(hebi_joint), j });
+    }
+  }
+
+  // Call derived class' load:
+  onLoad(model, sdf);
+
+  // Set up the update callbacks
+  this->update_connection_ = gazebo::event::Events::ConnectWorldUpdateBegin (
+    boost::bind(&HebiGazeboPlugin::onUpdateBase, this, _1));
```

Comment:
Does gazebo require boost here? If not, I would change this to something like

```c++
  this->update_connection_ = gazebo::event::Events::ConnectWorldUpdateBegin (
    [=](FunctionArgs... args) { HebiGazeboPlugin::onUpdateBase(this, args...); } );
```

Which isn't valid syntax, but I'm not sure what the arguments to that functon `onUpdateBase` are

Reply:
oops. The function is literally right below this...
Yeah, it looks like it'd be pretty easy to remove the boost call here, right? Unless `ConnectWorldUpdateBegin` requires it, that is...

Reply:
Unfortunately, the documentation for `ConnectWorldUpdateBegin` states: "Connect a `boost::slow` the the world update start signal." [sic]

http://osrf-distributions.s3.amazonaws.com/gazebo/api/2.2.1/classgazebo_1_1event_1_1Events.html#a441fb0fe08d924ab99b7255215e7502e